### DISCUSSION TITLE
Allow more flexibility to extension method definitions

### DISCUSSION THREAD

**Discussion by MithrilMan** [dotnet/csharplang#4511] [Language Ideas]
## Background and Motivation
Actually you can define an extension method only within a static, top level class.

To not pollute solution with a file/class that just holds extension methods that are meant to be used in a specific class, having a way to define such method within the class itself would help having a higher code management.

The behavior could be 
- extension method is accessible only in the class body (and eventually subclasses)
- extension method is visible everywhere within the namespace used to declare the class containing it

## Proposed API

I'd propose a keyword to be used to define an extension method within any class
let's call it `extension`

a definition of such method could include

```
extension void Apply(this Bar barInstance, int value){ }
extension Bar Apply(this Bar barInstance, int value){ }
extension T Apply<T>(this T barInstance, int value){ }
...
extension int Apply(this T barInstance, int value){ }
...
```

## Usage Examples

Suppose I've a class Foo and I want to be able to work (maybe in fluent like syntax like the example) on an instance of a Bar class, I'd like to be able to define my Foo class in a way similar to this

```
public class Foo {
   public MyMethod(Bar bar, int a, int b) {
      ...
      bar
         .Apply(a) // Bar class doesn't have a ApplyAB method, it's an extension defined below
         .Apply(b);
      ...
   }
   
   extension Bar Apply(this Bar barInstance, int value) {
     // do something with barInstance and value
     return barInstance;
   }
}
```

I don't mind about accessor or not (maybe it could be useful to hide it to inherited classes)

Feedback is welcome,
Thanks

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by TheUnlocked** [comment]
Similar proposals: #301, #799, #4366
Related: #1711
